### DISCUSSION TITLE
Pepuphysics as server with collisions best way to handle it

### DISCUSSION THREAD

**Discussion by Dendeii** [bepu/bepuphysics2#338] [Q&A] [answered]
Hi, i am building a space game bases on the planet example, the server is made in asp.net core and i have a unity client.
I use tcp / udp to sync the positions and that all works well.
But my problem is how to handle collisions for many objects i tried to used the ContactEvents example and that worked like i wanted but it started to breakdown when i add alot of asteroids for example.
I need the server to know about the collisions to correctly communicate it out to the clients that a collision was made.
What is the best performant option i can use in large scale simulations with 1000s of objects?

![image](https://github.com/user-attachments/assets/5fe95908-9a70-4899-926e-ced92d12485e)

**Comment by RossNordby** [comment, marked as answer]
This is a pretty broad question, so I doubt I'll be able to give you a real "best." One thing that comes to mind is that I wouldn't expect the approach in the contact events demo to struggle with performance at these scales. You could try something like `CollisionTrackingDemo` (and I do tend to prefer it over conventional events), but I don't actually expect the performance to differ by much. There's also `SolverContactEnumerationDemo`, but same thing.

My guess is the greater performance consideration is what is being done in response to the collisions, rather than the reporting of those collisions. If something expensive and/or requiring locks to synchronize was in the contact events, that could sequentialize execution and hurt performance a lot.

**Reply by Dendeii** [reply]
Thank you Ross!
I will test it more carefully and look at the CollisionTrackingDemo and SolverContactEnumerationDemo.
I suspect you might be right, i do some small processing in the event before offloading it to a seperate thread.
